File: android/src/main/java/de/paulweber/spenderino/android/screens/account/LoginView.kt
package de.paulweber.spenderino.android.screens.account

import androidx.compose.foundation.layout.Arrangement.Absolute.spacedBy
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.MaterialTheme
import androidx.compose.material.OutlinedTextField
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.focus.FocusDirection
import androidx.compose.ui.platform.LocalFocusManager
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import de.paulweber.spenderino.android.R
import de.paulweber.spenderino.android.views.PrimaryLoadingButton
import de.paulweber.spenderino.android.views.SecondaryLoadingButton
import de.paulweber.spenderino.viewmodel.AccountAction
import de.paulweber.spenderino.viewmodel.AccountState
import de.paulweber.spenderino.viewmodel.AccountViewModel

@Composable
fun LoginView(state: AccountState.Anonymous, viewModel: AccountViewModel) {
    Column(
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = spacedBy(12.dp)
    ) {
        val focusManager = LocalFocusManager.current
        val clearFocus = { focusManager.clearFocus() }
        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = spacedBy(8.dp)
        ) {
            EmailTextField(
                state,
                viewModel,
                keyboardActions = KeyboardActions(
                    onNext = { focusManager.moveFocus(FocusDirection.Down) }
                )
            )
            PasswordTextField(
                state,
                viewModel,
                keyboardActions = KeyboardActions(
                    onGo = {
                        clearFocus()
                        viewModel.perform(AccountAction.Login)
                    }
                )
            )
        }

        Spacer(modifier = Modifier.height(12.dp))

        LoginButton(state, viewModel, clearFocus)

        Text(
            stringResource(R.string.or),
            style = MaterialTheme.typography.body2,
        )

        RegisterButton(state, viewModel, clearFocus)
    }
}

@Composable
private fun LoginButton(
    state: AccountState.Anonymous,
    viewModel: AccountViewModel,
    clearFocus: () -> Unit
) {
    PrimaryLoadingButton(
        isLoading = state.isLoginLoading,
        onClick = {
            clearFocus()
            viewModel.perform(AccountAction.Login)
        },
        enabled = !(state.isLoginLoading || state.isRegisterLoading),
    ) { Text(stringResource(R.string.account_button_login)) }
}

@Composable
private fun RegisterButton(
    state: AccountState.Anonymous,
    viewModel: AccountViewModel,
    clearFocus: () -> Unit
) {
    SecondaryLoadingButton(
        enabled = !(state.isLoginLoading || state.isRegisterLoading),
        isLoading = state.isRegisterLoading,
        onClick = {
            clearFocus()
            viewModel.perform(AccountAction.Register)
        }
    ) { Text(stringResource(R.string.account_button_register)) }
}

@Composable
private fun EmailTextField(
    state: AccountState.Anonymous,
    viewModel: AccountViewModel,
    keyboardActions: KeyboardActions
) {
    OutlinedTextField(
        value = state.email,
        onValueChange = { viewModel.perform(AccountAction.ChangeEmailText(it)) },
        keyboardOptions = KeyboardOptions(
            keyboardType = KeyboardType.Email,
            imeAction = ImeAction.Next
        ),
        keyboardActions = keyboardActions,
        label = { Text(stringResource(R.string.account_email_label)) },
        placeholder = { Text(stringResource(R.string.account_email_hint)) },
        enabled = !(state.isLoginLoading || state.isRegisterLoading),
        maxLines = 1,
    )
}

@Composable
private fun PasswordTextField(
    state: AccountState.Anonymous,
    viewModel: AccountViewModel,
    keyboardActions: KeyboardActions
) {
    OutlinedTextField(
        value = state.password,
        onValueChange = { viewModel.perform(AccountAction.ChangePasswordText(it)) },
        visualTransformation = PasswordVisualTransformation(),
        keyboardOptions = KeyboardOptions(
            keyboardType = KeyboardType.Password,
            imeAction = ImeAction.Go
        ),
        keyboardActions = keyboardActions,
        label = { Text(stringResource(R.string.account_password_label)) },
        placeholder = { Text(stringResource(R.string.account_password_hint)) },
        enabled = !(state.isLoginLoading || state.isRegisterLoading),
        maxLines = 1,
    )
}

@Preview(showBackground = true)
@Composable
private fun Preview() {
    LoginView(
        state = AccountState.Anonymous(),
        viewModel = AccountViewModel()
    )
}

File: ios/Spenderino/Source/Screens/Account/LoginView.swift
import SwiftUI
import shared

struct LoginView: View {
    let viewModel: AccountViewModel
    let state: AccountState.Anonymous

    var body: some View {
        VStack(spacing: 32) {
            VStack(alignment: .leading, spacing: 4) {
                Text("account_email_label")
                    .font(.caption)
                    .fontWeight(.semibold)

                emailTextField

                Spacer().frame(height: 4)

                Text("account_password_label")
                    .font(.caption)
                    .fontWeight(.semibold)

                passwordTextField
            }

            VStack(alignment: .center) {
                loginButton
                Text("or")
                    .font(.footnote)
                    .foregroundColor(.secondary)
                registerButton
            }
        }
        .padding()
        .disabled(state.isRegisterLoading || state.isLoginLoading)
    }

    var loginButton: some View {
        LoadingButton(isLoading: state.isLoginLoading) {
            viewModel.perform(action: AccountAction.Login())
        } label: {
            Text("account_button_login")
        }
        .buttonStyle(.borderedProminent)
    }

    var registerButton: some View {
        LoadingButton(isLoading: state.isRegisterLoading) {
            viewModel.perform(action: AccountAction.Register())
        } label: {
            Text("account_button_register")
        }
        .buttonStyle(.bordered)
    }

    var emailTextField: some View {
        TextField(
            "account_email_hint",
            text: Binding<String>(
                get: { state.email },
                set: { viewModel.perform(action: AccountAction.ChangeEmailText(text: $0)) }
            )
        )

        .textFieldStyle(.roundedBorder)
        .textContentType(.emailAddress)
        .autocapitalization(.none)
    }

    var passwordTextField: some View {
        SecureField(
            "account_password_hint",
            text: Binding<String>(
                get: { state.password },
                set: { viewModel.perform(action: AccountAction.ChangePasswordText(text: $0)) }
            )
        )
        .textFieldStyle(.roundedBorder)
        .textContentType(.password)
    }
}

struct AnonymousAccountView_Previews: PreviewProvider {
    static var previews: some View {
        let viewModel = AccountViewModel(
            state: AccountState.Anonymous(
                isLoginLoading: true,
                isRegisterLoading: true,
                email: "",
                password: ""
            ),
            route: nil,
            onSetupComplete: {},
            onBack: {}
        )
        LoginView(viewModel: viewModel, state: .init(
            isLoginLoading: false,
            isRegisterLoading: false,
            email: "Test",
            password: "GANZ VIEL STUFF")
        ).previewLayout(.sizeThatFits)
    }
}
